### PR TITLE
py/modmath: Add math.tau, math.nan and math.inf constants.

### PR DESCRIPTION
**Pull request by stinos**
These have been in Python for a while now so for CPython compatibility it's a good thing to have them. I used MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_FULL_FEATURES because, well, I figured since we have it might as well use it. Or esle use it as discussion to make clear how it should be used :)

### CODE REVIEW

**Comment by dpgeorge** [comment]
I don't think this feature will build on unix nanbox, nor esp8266 (object reprs C and D).

**Comment by stinos** [comment]
Can you elaborate? I mean, no it won't build because it'll be excluded (sliently) unless using obj repr A or B, just like e and pi. Do you mean there should ba en `#error` when trying to put MICROPY_PY_MATH_CONSTANTS to 1? Or is there something which I overlooked leading to an error already for e.g. nanbox?

**Comment by dpgeorge** [comment]
Ah, you're right, it will build.  But I think it's misleading because enabling that feature won't do anything with obj repr C or D.

File: py/modmath.c [hunk 1]
```diff
@@ -371,6 +371,15 @@ STATIC const mp_rom_map_elem_t mp_module_math_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_math) },
     { MP_ROM_QSTR(MP_QSTR_e), mp_const_float_e },
     { MP_ROM_QSTR(MP_QSTR_pi), mp_const_float_pi },
+    #ifdef mp_const_float_tau
```
Comment (dpgeorge) [inline py/modmath.c:374]: IMO this should be `#if MICROPY_PY_MATH_CONSTANTS`, to match how other options are included/excluded.

Comment (dpgeorge) [reply py/modmath.c:374]: Then of course these values need to be defined for obj repr C and D.

**Comment by dpgeorge** [comment]
> These have been in Python for a while now so for CPython compatibility it's a good thing to have them

The points in favour of adding them would be: 1) CPython compat, 2) they are useful.

The point against would be that they can be obtained other ways, eg `float('nan')`, `2 * math.pi`, so are not necessary for a "micro" Python.

Again this is the usual debate of Micro vs Python.

> I used MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_FULL_FEATURES because, well, I figured since we have it might as well use it. Or else use it as discussion to make clear how it should be used :)

Yes, these feature levels were added to help make it easier to add features in a controlled way.  But seeing how all other optional `math` features are controlled by `MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_EXTRA_FEATURES`, maybe it makes sense to use that level?   That will make it available on rp2 and stm32...

File: py/modmath.c (hunk 1 above)
Comment (stinos) [reply py/modmath.c:374]: Ok

**Comment by stinos** [comment]
> Again this is the usual debate of Micro vs Python.

yeah and as usual it's really hard to tell whether this is a welcome feature for multiple people :) That's also a reason I chose the 'full' features; but indeed since the rest of math uses 'extra' this should as well.

I added tests to make sure things are ok, but currently no tests have both `MICROPY_PY_MATH_CONSTANTS` and obj repr C or D, whereas that's what mainly needs testing now. Suggestion for where to add that?

The tests fail for qemu-arm and I don't know why; perhaps asked this already but could we have the CI print the actual failures?

The cmath modules has similar constants, does that also need to be changed with a similar `MICROPY_PY_CMATH_CONSTANTS` ?

**Comment by dpgeorge** [comment]
> I added tests to make sure things are ok, but currently no tests have both `MICROPY_PY_MATH_CONSTANTS` and obj repr C or D, whereas that's what mainly needs testing now. Suggestion for where to add that?

Would make sense to enable this feature on unix-nanbox builds, which would then be tested in CI.  That would do obj repr D.  For repr C... that would need a new unix or qemu-arm build in CI to test that.  Probably a good idea to add such a build but not in this PR.

> The tests fail for qemu-arm and I don't know why; perhaps asked this already but could we have the CI print the actual failures?

It prints out which test file failed, and in this case it's `math_constants.py`.  Might need to add `-C 100` to the grep in `.github/workflows/ports_qemu-arm.yml` to find out more info about that failure (can do that in this PR as a new commit).

> The cmath modules has similar constants, does that also need to be changed with a similar `MICROPY_PY_CMATH_CONSTANTS` ?

For consistency it would make sense, but I really don't see the usefulness of having these constants duplicated in `cmath` (maybe there's a good reason?).  So I suggest leaving them out of there for now.

**Comment by dpgeorge** [comment]
> Ok, just in the same commit where the constants get introduced right?

Yes.

> The `math.pi == 2.0 * asin(1.0)` test fails.

Hmm, not sure what FP guarantees there are about `asin(1.0)` being exactly pi/2, or 2.0 * pi/2 being exactly pi... it could be showing a problem somewhere in the qemu-arm port.

For now I guess just add a small tolerance to that test.

**Comment by stinos** [comment]
Changed the test to `print(cos(math.pi))` which should print `-1.0` everywhere.

About the qemu-arm test failure printing: it now greps for `"FAIL"` which matches both test failures (`[%s FAILED]`) and the summary (`%d/%d TESTS FAILED. (%d skipped)`). But what comes beofre that last one is irrelevant since it's just the output of the last tests which not necessarily failed. Would it be ok to grep for `"FAILED\]"` instead?

**Comment by dpgeorge** [comment]
> But what comes beofre that last one is irrelevant since it's just the output of the last tests which not necessarily failed. Would it be ok to grep for `"FAILED\]"` instead?

I think the more output the better (within reason, it's too much to just dump that entire file).  Remember this output is only printed on failure, and if it fails you really want to know why.  So as you have it now looks fine.

**Comment by dpgeorge** [comment]
Merged in a9448c0a86d6d1f04e190d5c093ca93d9470b99a through f3229590a9ce8286627f5763884328cdc970fa1d

Nice and clean, thanks!

**Pull request closed without merging**